packages/dht/src/dht/SortedContactList.ts

```diff
@@ -0,0 +1,132 @@
+import KBucket from 'k-bucket'
+import { PeerID } from '../helpers/PeerID'
+import { DhtPeer } from './DhtPeer'
+
+class ContactWrapper {
+    public contacted = false
+    public active = false
+    constructor(public contact: DhtPeer) {
+    }
+}
+
+export class SortedContactList {
+    private contactsById: { [id: string]: ContactWrapper } = {}
+    private contactIds: PeerID[] = []
+
+    constructor(private ownId: PeerID, private maxSize: number) {
+        this.compareIds = this.compareIds.bind(this)
+        this.ownId = ownId
+    }
+
+    public getClosestContactId(): PeerID {
+        return this.contactIds[0]
+    }
+
+    public getContactIds(): PeerID[] {
+        return this.contactIds
+    }
+
+    public addContact(contact: DhtPeer): void {
+        if (this.ownId.equals(contact.peerId)) {
+            return
+        }
+        if (!this.contactsById.hasOwnProperty(contact.peerId.toMapKey())) {
+            if (this.contactIds.length < this.maxSize) {
+                this.contactsById[contact.peerId.toMapKey()] = new ContactWrapper(contact)
+                this.contactIds.push(contact.peerId)
+                this.contactIds.sort(this.compareIds)
+            
+            } else if (this.compareIds(this.contactIds[this.maxSize - 1], contact.peerId) > 0) {
+                const removed = this.contactIds.pop()
+                delete this.contactsById[removed!.toMapKey()]
+                this.contactsById[contact.peerId.toMapKey()] = new ContactWrapper(contact)
+                this.contactIds.push(contact.peerId)
+                this.contactIds.sort(this.compareIds)
+            }
+        }  
+    }
+
+    public addContacts(contacts: DhtPeer[]): void {
+        contacts.forEach( (contact) => this.addContact(contact))
+    }
+
+    public setContacted(contactId: PeerID): void {
+        if (this.contactsById.hasOwnProperty(contactId.toMapKey())) {
+            this.contactsById[contactId.toMapKey()].contacted = true
+        }
+    }
+
+    public setActive(contactId: PeerID): void {
+        if (this.contactsById.hasOwnProperty(contactId.toMapKey())) {
+            this.contactsById[contactId.toMapKey()].active = true
+        }
+    }
+
+    public getUncontactedContacts(num: number): DhtPeer[] {
```

Comment:
Is this actually a `NodeList` or `NeighborList`? If it is a list of contacts, I'd guess from the name that each member has already been contacted.

Reply:
Contacted is set during the join operation when getClosestPeers is requested on the peer. It is an important part of closing the join() operation.

Reply:
Contact here refers to a peer that you have heard of. You may have contacted it and you may not have contacted it. You may have heard of the peer via another contact or by them directly sending you a message or by seeing a routed message originated by them.

The SortedContactList class could be seen as a "phonebook"